### PR TITLE
Make sabakan integration better

### PR DESCRIPTION
**Pull request by kfyharukz**
This PR makes Sabakan integration better by the new scoring & filtering methods.
By the new methods, CKE  distributes Nodes over the racks as even as possible.

### CODE REVIEW

File: sabakan/generator.go [hunk 1]
```diff
@@ -285,11 +257,47 @@ func (g *Generator) SetWaitSeconds(secs float64) {
 	g.waitSeconds = secs
 }
 
+// selectControlPlaneFromUnused selects a healthy controle plane from unused or worker machines.
```
Comment (kfyharukz) [inline sabakan/generator.go:260]: the func name in the comment is wrong.

File: sabakan/generator.go [hunk 2]
```diff
@@ -285,11 +257,47 @@ func (g *Generator) SetWaitSeconds(secs float64) {
...
+	}
+
+	numCPsByRack := g.countMachinesByRack(true, "")
+	sort.Slice(candidates, func(i, j int) bool {
+		si := scoreMachine(candidates[i], numCPsByRack, g.timestamp)
+		sj := scoreMachine(candidates[j], numCPsByRack, g.timestamp)
+		// higher first
+		return si > sj
+	})
+
```
Comment (kfyharukz) [inline sabakan/generator.go:280]: remove the line.

File: sabakan/generator_test.go [hunk 3]
```diff
@@ -1118,18 +1119,320 @@ func TestCountWorkersByRack(t *testing.T) {
...
+		}
+	})
+
+	t.Run("IncreaseMinimumWorkers", func(t *testing.T) {
+		rackIDs := []int{0, 1, 2}
+		var machines []Machine
+		for i := range rackIDs {
+			machines = append(machines, createRack(i, "standard")...)
+		}
+		constraints := &cke.Constraints{
```
Comment (kfyharukz) [inline sabakan/generator_test.go:1194]: move `constraints` to the near of use lines.

File: sabakan/score.go [hunk 4]
```diff
@@ -5,45 +5,69 @@ import (
 )
 
 const (
-	baseScore      = 10
-	unhealthyScore = -3 * baseScore
+	// maxCountPerRack should be max count of machines per rack + 1
```
Comment (kfyharukz) [inline sabakan/score.go:8]: the expression "more than" is appropriate.

File: docs/sabakan-integration.md [hunk 5]
```diff
@@ -179,24 +180,27 @@ Algorithms
...
-    - If the machine's lifetime is > 1000 days, +10 (+30 in total).
-    - If the machine's lifetime is < -250 days, -10.
-    - If the machine's lifetime is < -500 days, -10 (-20 in total).
-    - If the machine's lifetime is < -1000 days, -10 (-30 in total).
-    - If the cluster contains `n` machines in the same rack, - `min(n, 10)`.
-5. Select the highest scored machine.
+4. Add the following score to each machine:
+    - (100 - (machine counts which have the same role and in the same rack)) * 100
+5. Add the following score to each machine:
+    - If the machine's state is healthy, +10.
```
[the remark this answers is not in the document]
Comment (ymmt2005) [reply docs/sabakan-integration.md:195]: Also, you missed the below condition...

> - If the machine is not healthy, -30.

File: docs/sabakan-integration.md [hunk 6]
```diff
@@ -283,7 +287,7 @@ nodes are added.
 
 #### Decrease worker nodes
 
-If a worker node is either retiring or retired for a while,
+If a worker node is retired for a while,
```
Comment (ymmt2005) [inline docs/sabakan-integration.md:290]:
```suggestion
If a worker node is kept retired for a while,
```

File: sabakan/score.go [hunk 7]
```diff
@@ -5,45 +5,60 @@ import (
 )
 
 const (
-	baseScore      = 10
-	unhealthyScore = -3 * baseScore
+	// maxCountPerRack should be more than max machine num per rack + 1
+	maxCountPerRack = 100
+	// heathyScore is addted when the machine status is healthy.
+	heathyScore = 1000
```
Comment (ymmt2005) [inline sabakan/score.go:11]: healthy?

File: sabakan/score.go [hunk 8]
```diff
@@ -5,45 +5,60 @@ import (
...
+	return heathyScore + score
+}
 
-	return score
+func filterHealthyMachinesByRole(ms []*Machine, role string) []*Machine {
+	var filtered []*Machine
+	for _, m := range ms {
+		if m.Status.State != StateHealthy || (role != "" && m.Spec.Role != role) {
+			continue
+		}
```
Comment (ymmt2005) [inline sabakan/score.go:59]: I'd like to suggest a bit refactor for better readability.

```suggestion
		if m.Status.State != StateHealthy {
			continue
		}
		if role != "" && m.Spec.Role != role {
			continue
		}
```

File: sabakan/generator.go [hunk 9]
```diff
@@ -119,13 +111,12 @@ type Generator struct {
 	nextUnused        []*Machine
 	nextControlPlanes []*Machine
 	nextWorkers       []*Machine
-	workersByRole     map[string]int
 }
 
 // NewGenerator creates a new Generator.
 // current can be nil if no cluster configuration has been set.
```
Comment (ymmt2005) [inline sabakan/generator.go:117]: `current` no longer exists.

File: sabakan/generator.go [hunk 10]
```diff
@@ -234,108 +195,104 @@ func (g *Generator) getWorkerTmpl(role string) nodeTemplate {
...
+func removeMachine(machines []*Machine, m *Machine) []*Machine {
+	var ret []*Machine
+	for i, mm := range machines {
+		if m == mm {
+			ret = append(machines[:i], machines[i+1:]...)
+			break
 		}
 	}
-	if len(unused) == 0 {
+	return ret
```
Comment (ymmt2005) [inline sabakan/generator.go:211]: If `m` is not found in `machines`, this function returns nil.
Is this intended?  If so, please add a comment to warn users.

File: sabakan/generator.go [hunk 11]
```diff
@@ -195,20 +195,20 @@ func (g *Generator) getWorkerTmpl(role string) nodeTemplate {
...
 }
 
-// SetWaitSeconds set seconds before removing retiring nodes from the cluster.
+// SetWaitSeconds set seconds before removing retired nodes from the cluster.
 func (g *Generator) SetWaitSeconds(secs float64) {
 	g.waitSeconds = secs
 }
 
+// removeMachine removes the specified machine from machines slice.
+// If m is nil, this returns the same slice as input.
```
Comment (ymmt2005) [inline sabakan/generator.go:204]: and if `m` is not found in `machines`?

File: sabakan/generator.go [hunk 12]
```diff
@@ -195,20 +195,20 @@ func (g *Generator) getWorkerTmpl(role string) nodeTemplate {
...
-	var ret []*Machine
 	for i, mm := range machines {
 		if m == mm {
-			ret = append(machines[:i], machines[i+1:]...)
-			break
+			return append(machines[:i], machines[i+1:]...)
 		}
 	}
-	return ret
+	return machines
```
Comment (ymmt2005) [inline sabakan/generator.go:211]: I think `panic` is appropriate if `m` must be in `machines` .

Comment (masa213f) [reply sabakan/generator.go]: Thanks. I fixed so.

File: sabakan/generator.go (hunk 9 above)
Comment (ymmt2005) [reply sabakan/generator.go:117]: @masa213f This is not fixed yet.

**Pull request merged**